lib/phoenix/socket/transport.ex

```diff
@@ -388,6 +388,50 @@ defmodule Phoenix.Socket.Transport do
     end
   end
 
+  @doc false
+  def extract_metadata(conn, keys) do
+    %{}
+    |> extract_peer_data(conn, Enum.member?(keys, :peer_data))
+    |> extract_x_headers(conn, Enum.member?(keys, :x_headers))
+    |> extract_uri(conn, Enum.member?(keys, :uri))
```

Comment:
This is not very efficient. It would be better to call a `Enum.reduce/3` on the keys. It would then also be a good opportunity to raise on any unknown keys.